File: app/src/main/java/by/tuzau/andrey/rusandroidapp/ui/adapter/FileSystemArrayAdapter.java
package by.tuzau.andrey.rusandroidapp.ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.List;

import by.tuzau.andrey.rusandroidapp.R;
import by.tuzau.andrey.rusandroidapp.data.domain.AbstractFileSystemElement;
import by.tuzau.andrey.rusandroidapp.data.domain.FileElement;
import by.tuzau.andrey.rusandroidapp.data.domain.FolderElement;
import by.tuzau.andrey.rusandroidapp.data.domain.TypeElement;
import by.tuzau.andrey.rusandroidapp.ui.adapter.viewholder.FileViewHolder;
import by.tuzau.andrey.rusandroidapp.ui.adapter.viewholder.FolderViewHolder;

/**
 * Created by User on 19.11.2017.
 */
public class FileSystemArrayAdapter extends BaseAdapter {

    /** Count of view's type */
    private static final int VIEWTYPE_COUNT = 2;
    /** List of adapter's data */
    private List<AbstractFileSystemElement> data;
    /** Object for creating view from .xml file */
    private final LayoutInflater inflater;

    private final OnClickFolderListener listener;

    public interface OnClickFolderListener {
        void onClickFolder(AbstractFileSystemElement element);
    }

    public FileSystemArrayAdapter(Context context, FolderElement folderElement, OnClickFolderListener listener) {
        this.data = folderElement.getListFileSystemElement();
        this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.listener = listener;
    }

    @Override
    public int getViewTypeCount() {
        return VIEWTYPE_COUNT;
    }

    @Override
    public int getItemViewType(int position) {
        return data.get(position).getTypeElement().getIndex();
    }

    @Override
    public int getCount() {
        return data.size();
    }

    @Override
    public Object getItem(int position) {
        return data.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(final int position, View view, ViewGroup viewGroup) {
        TypeElement typeElement = data.get(position).getTypeElement();
        if (view == null) {
            // Creating view from .xml file depending on typeElement
            switch (typeElement) {
                case FILE:
                    view = inflater.inflate(R.layout.file_element, null);
                    FileViewHolder fileViewHolder = new FileViewHolder();
                    fileViewHolder.tvFileName = (TextView) view.findViewById(R.id.tvFileName);
                    view.setTag(fileViewHolder);
                    break;
                case FOLDER:
                    view = inflater.inflate(R.layout.folder_element, null);
                    FolderViewHolder folderViewHolder = new FolderViewHolder();
                    folderViewHolder.tvFileName = (TextView) view.findViewById(R.id.tvFolderName);
                    folderViewHolder.tvCountAudioFiles = (TextView) view.findViewById(R.id.tvCountAudioFiles);
                    view.setTag(folderViewHolder);
                    break;
            }
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onClickFolder(data.get(position));
                }
            });
        }
        // Setting data values
        switch (typeElement) {
            case FILE:
                FileViewHolder fileViewHolder = (FileViewHolder) view.getTag();
                FileElement file = (FileElement) data.get(position);
                fileViewHolder.tvFileName.setText(file.getName());
                break;
            case FOLDER:
                FolderViewHolder folderViewHolder = (FolderViewHolder) view.getTag();
                FolderElement folder = (FolderElement) data.get(position);
                folderViewHolder.tvFileName.setText(folder.getName());
                folderViewHolder.tvCountAudioFiles.setText("Count audio: " + folder.getCountAudioFiles());
                break;
        }
        return view;
    }
}

File: app/src/main/java/by/tuzau/andrey/rusandroidapp/ui/adapter/FileSystemArrayAdapter.kt
package by.tuzau.andrey.rusandroidapp.ui.adapter

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import by.tuzau.andrey.rusandroidapp.R
import by.tuzau.andrey.rusandroidapp.data.domain.AbstractFileSystemElement
import by.tuzau.andrey.rusandroidapp.data.domain.FolderElement
import by.tuzau.andrey.rusandroidapp.data.domain.TypeElement
import by.tuzau.andrey.rusandroidapp.ui.adapter.viewholder.FileViewHolder
import by.tuzau.andrey.rusandroidapp.ui.adapter.viewholder.FolderViewHolder

class FileSystemArrayAdapter(
        context: Context,
        folderElement: FolderElement,
        private val listener: OnClickFolderListener
) : BaseAdapter() {

    /** Count of view's type */
    private val VIEWTYPE_COUNT = 2
    /** List of adapter's data */
    private val data: List<AbstractFileSystemElement> = folderElement.listElement
    /** Object for creating view from .xml file */
    private var inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    interface OnClickFolderListener {
        fun onClickFolder(element: AbstractFileSystemElement)
    }

    override fun getViewTypeCount() = VIEWTYPE_COUNT

    override fun getItemViewType(position: Int) = data[position].typeElement.index

    override fun getCount(): Int = data.size

    override fun getItem(position: Int) = data[position]

    override fun getItemId(position: Int) = position.toLong()

    override fun getView(position: Int, view: View?, viewGroup: ViewGroup?): View {
        val typeElement = data[position].typeElement
        var currentView: View? = view
        if (view == null) {
            // Creating view from .xml file depending on typeElement
            when (typeElement) {
                TypeElement.FILE -> {
                    currentView = inflater.inflate(R.layout.file_element, null)
                    val fileViewHolder = FileViewHolder()
                    fileViewHolder.tvFileName = currentView.findViewById<TextView>(R.id.tvFileName)
                    currentView.tag = fileViewHolder
                }
                TypeElement.FOLDER -> {
                    currentView = inflater.inflate(R.layout.folder_element, null);
                    val folderViewHolder = FolderViewHolder()
                    folderViewHolder.tvFileName = currentView.findViewById<TextView>(R.id.tvFolderName)
                    folderViewHolder.tvCountAudioFiles = currentView.findViewById<TextView>(R.id.tvCountAudioFiles)
                    currentView.tag = folderViewHolder
                    currentView.setOnClickListener({_ -> listener.onClickFolder(data[position])})
                }
            }
        }
        // Setting data values
        when (typeElement) {
            TypeElement.FILE -> {
                val fileViewHolder: FileViewHolder = currentView!!.tag as FileViewHolder
                val file = data[position]
                fileViewHolder.tvFileName.text = file.name
                Log.d("FileSystemArrayAdapter", file.name)
            }
            TypeElement.FOLDER -> {
                val folderViewHolder: FolderViewHolder = currentView!!.tag as FolderViewHolder
                val folder = data!![position] as FolderElement
                folderViewHolder.tvFileName.text = folder.name
                folderViewHolder.tvCountAudioFiles.text = "Count audio: ${folder.countAudioFiles}"
            }
        }
        return currentView!!;
    }

}